tests: kernel: interrupt: make test work with any available NVIC IRQ

ioannisg opened the pull request:
This commit re-works the test for the ARM architecture,
so that it can work with any available NVIC IRQ, not
bound to use the last 2 NVIC lines. It makes use of
the dynamic IRQ feature.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

Fixes #18654

ioannisg committed:
tests: kernel: interrupt: make test work with any available NVIC IRQ

This commit re-works the test for the ARM architecture,
so that it can work with any available NVIC IRQ, not
bound to use the last 2 NVIC lines. It makes use of
the dynamic IRQ feature.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@galak can you check if this fixes #18654 ??

galak reviewed (changes requested):
Can you update tests/kernel/interrupt/testcase.yaml and remove all the `platform_excludes`

```
diff --git a/tests/kernel/interrupt/testcase.yaml b/tests/kernel/interrupt/testcase.yaml
index c061cef483..e1733cd79e 100644
--- a/tests/kernel/interrupt/testcase.yaml
+++ b/tests/kernel/interrupt/testcase.yaml
@@ -1,9 +1,6 @@
 tests:
   arch.interrupt:
     arch_exclude: nios2 riscv32 arc qemu_x86 qemu_x86_coverage qemu_x86_64
-    platform_exclude: hexiwear_kw40z frdm_kw41z frdm_kl25z nucleo_f103rb
-        nucleo_f091rc olimexino_stm32 usb_kw24d512 stm32_min_dev_blue
-        stm32_min_dev_black v2m_beetle
     tags: interrupt
 ```

galak reviewed (commented):
General concepts look good.

ioannisg commented on tests/kernel/interrupt/src/interrupt.h:
```diff
@@ -11,6 +11,43 @@
 #if defined(CONFIG_ARM)
 #include <arch/arm/cortex_m/cmsis.h>
+
+static u32_t irq_line_0;
```
Nope, `irq_line_1` needs to be visible by `handler()`, which enables the irq1.

ioannisg replied:
They could be defined on top of nested_irq.c, alternatively; I added them here, to skip one more #ifdef CONFIG_ARM block..

ioannisg commented on tests/kernel/interrupt/src/nested_irq.c:
```diff
@@ -82,13 +87,29 @@ void isr0(void *param)
+	printk("IRQS %u %u \n", irq_line_0, irq_line_1);
+	z_arch_irq_connect_dynamic(irq_line_0,
+		ISR0_PRIO, isr0, NULL, 0);
+	z_arch_irq_connect_dynamic(irq_line_1,
```
Fixed

ioannisg commented on tests/kernel/interrupt/src/nested_irq.c:
```diff
@@ -82,13 +87,29 @@ void isr0(void *param)
+	irq_line_0 = get_available_nvic_line(CONFIG_NUM_IRQS);
+	irq_line_1 = get_available_nvic_line(irq_line_0);
+	printk("IRQS %u %u \n", irq_line_0, irq_line_1);
+	z_arch_irq_connect_dynamic(irq_line_0,
```
Fixed

ioannisg committed:
tests: kernel: interrupt: stop excluding ARM platforms

As we have re-worked the test code, and the test-case can run
on Cortex-M platforms on any available and implemented NVIC
IRQ lines, we do not need to exclude these ARM boards anymore.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
> 
> 
> Can you update tests/kernel/interrupt/testcase.yaml and remove all the `platform_excludes`
> 
> ```
> diff --git a/tests/kernel/interrupt/testcase.yaml b/tests/kernel/interrupt/testcase.yaml
> index c061cef483..e1733cd79e 100644
> --- a/tests/kernel/interrupt/testcase.yaml
> +++ b/tests/kernel/interrupt/testcase.yaml
> @@ -1,9 +1,6 @@
>  tests:
>    arch.interrupt:
>      arch_exclude: nios2 riscv32 arc qemu_x86 qemu_x86_coverage qemu_x86_64
> -    platform_exclude: hexiwear_kw40z frdm_kw41z frdm_kl25z nucleo_f103rb
> -        nucleo_f091rc olimexino_stm32 usb_kw24d512 stm32_min_dev_blue
> -        stm32_min_dev_black v2m_beetle
>      tags: interrupt
> ```

Done

ioannisg commented:
@galak please, revisit this, so it can be added for v2.0.0

galak reviewed (commented):
Looks good.  Only change I'd ask is to move the static u32_t irq_line_{0,1} into the .c file.

ioannisg commented:
> 
> 
> Looks good. Only change I'd ask is to move the static u32_t irq_line_{0,1} into the .c file.

Sure, I will do this.

ioannisg commented:
Tested on the following ARM platforms:
- atsamr21_xpro
- sam_e70_xplained
- nrf9160_pca10090
- nrf52840_pca10056
- nrf51_pca10028
- mimxrt1015_evk
- frdm_k64f
- lpcxpress054114_m4
- nrf52_pca10040

ioannisg commented:
Verified, also, that it builds for the platforms that are, now, not excluded.

ioannisg commented on tests/kernel/interrupt/testcase.yaml:
```diff
@@ -1,9 +1,6 @@
 tests:
   arch.interrupt:
     arch_exclude: nios2 riscv32 arc qemu_x86 qemu_x86_coverage qemu_x86_64
```
will remove :)

ioannisg replied:
Removed. I kept the excluded platforms, though.
